### PR TITLE
Docker Support with GPU

### PR DESCRIPTION
**Pull request by binarygeek119**
i have created a dockerfile that does build a image that will work with a GPU pass thought
 to run this docker you need to run this command
`docker build -t ultrasinger .`

then after the build run.

`docker run --gpus all -it --name UltraSinger ultrasinger`

right now this drops you into the docker commandline.
all you have to do is run `python3 UltraSinger.py -i file`.

right now i edited the dockerfile to just copy a mp3 for testing and all works fine for me.
rightnow to use this you must run this from commandline. but its all self contained.
to get this in and out run this command instead.
`docker run --gpus all -it --name UltraSinger -v  $(pwd)/mp3:/mp3 ultrasinger` 
this will let you get files in and out of the docker.

the next step would be more testing and then making a webui to pass eveything though.

im sure there more needed then what i came up with, this is my first docker ive build.

thoughts?

### CODE REVIEW

**Comment by binarygeek119** [comment]
While this does produce a file that is useable, it doesn't use the GPU for everything. Some dependency not ment In the docker. Still playing with it. I may have a idea what's wrong but it's late and I need to go to bed. Will keep working on it.

**Comment by rakuri255** [comment]
> While this does produce a file that is useable, it doesn't use the GPU for everything. Some dependency not ment In the docker. Still playing with it. I may have a idea what's wrong but it's late and I need to go to bed. Will keep working on it.

Cool! will try it. 
You can upgrade pytorch and Tensorflow. 
The Tensorflow updgrade was needed for Colab, since they used Cuda 12.1.
Maybe its a similar issue?

```
!pip3 install torch==2.3.1+cu121 torchvision==0.18.1+cu121 torchaudio==2.3.1+cu121 --index-url https://download.pytorch.org/whl/cu121
!pip install tensorflow[and-cuda]==2.15.0
```

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+WORKDIR /app
+
+COPY requirements.txt /app
+RUN pip install --no-cache-dir -r requirements.txt
+RUN git clone https://github.com/rakuri255/UltraSinger.git
+WORKDIR /app/UltraSinger
+RUN cp -Rv * ..
+RUN apt install ffmpeg curl -y
+RUN curl -L https://github.com/yt-dlp/yt-dlp/releases/latest/download/yt-dlp -o /app/yt-dlp
```
[the remark this answers is not in the document]
Comment (binarygeek119) [reply Dockerfile]: this was left over from debugging. thought it was a issues on docker. removed it.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,25 @@
+FROM nvidia/cuda:12.5.0-runtime-ubuntu22.04
+
+WORKDIR /app
+RUN apt update
+RUN apt install git python3-pip -y
+RUN pip3 install torch==2.0.1+cu117 torchvision==0.15.2+cu117 torchaudio==2.0.2+cu117 --index-url https://download.pytorch.org/whl/cu117
```
[the remark this answers is not in the document]
Comment (binarygeek119) [reply Dockerfile]: i had to do a bit of tweaking to the line but it now works.

**Comment by binarygeek119** [comment]
i have all steps working on gpu now never seen it go so fast. it does take a bit to build. do you want me to upload to docker for people to pull from? i do know there are ways you can link github to rebuild the docker when i update come out. for this i have no idea how to do it. but i would be happy to upload a new docker image when updates come out.

**Comment by binarygeek119** [comment]
i have tested it about 5 times now and it seems to be put together right. i need other to test to see any issues.

**Comment by rakuri255** [comment]
> i have all steps working on gpu now never seen it go so fast. it does take a bit to build. do you want me to upload to docker for people to pull from? i do know there are ways you can link github to rebuild the docker when i update come out. for this i have no idea how to do it. but i would be happy to upload a new docker image when updates come out.

Nice! Good work! We need an github CI for this. Its not that hard -> [publishing-docker-images](https://docs.github.com/en/actions/publishing-packages/publishing-docker-images)
We can discusse in discord about that and try it out.

Can you describe the docker changes a bit in the readme? Then i accept the PR.

**Comment by rakuri255** [comment]
Check [CodeFactor — 7 issues found](https://www.codefactor.io/repository/github/rakuri255/ultrasinger/pull/148).

**Comment by binarygeek119** [comment]
testing the updated dockerfile with the fixes

**Comment by binarygeek119** [comment]
i think i got it now just have to build

**Comment by binarygeek119** [comment]
it fully works and no errors.

**Review by rakuri255 — APPROVED** [review]

**Pull request merged**